Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md

```diff
@@ -785,31 +785,55 @@ You can detect popular reverse engineering tools that have been installed in an
-```c
+With API Level 25 and below it was possible to query for all running services by using the Java method  (getRunningServices[https://developer.android.com/reference/android/app/ActivityManager.html#getRunningServices(int) "getRunningServices"]. This allows to iterate through the list of running UI activities, but will not show you daemons like the frida-server. Starting with API Level 26 and above `getRunningServices()` will even only return the caller's own services.
+
+A working solution to detect the frida process is to us the command `ps` instead.
+
+```Java
 public boolean checkRunningProcesses() {
 
-  boolean returnValue = false;
+    boolean returnValue = false;
+
+    try {
+        Process process = Runtime.getRuntime().exec("ps");
+        BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
+        int read;
+        char[] buffer = new char[4096];
+        StringBuffer output = new StringBuffer();
+        while ((read = reader.read(buffer)) > 0) {
+            output.append(buffer, 0, read);
+        }
+        reader.close();
 
-  // Get currently running application processes
-  List<RunningServiceInfo> list = manager.getRunningServices(300);
+        // Waits for the command to finish.
+        process.waitFor();
+        Log.d("fridaserver", output.toString());
 
-  if(list != null){
-    String tempName;
-    for(int i=0;i<list.size();++i){
-      tempName = list.get(i).process;
+        if(output.toString().contains("frida-server")) {
+            Log.d("fridaserver","Frida Server process found!" );
+            returnValue = true;
+        }
+
+    } catch (IOException e) {
+
+    } catch (InterruptedException e) {
 
-      if(tempName.contains("fridaserver")) {
-        returnValue = true;
-      }
     }
-  }
-  return returnValue;
+
+    return returnValue;
 }
+
 ```
 
-This works if Frida is run in its default configuration. Perhaps it's also enough to stump some script kiddies during their first steps in reverse engineering. It can, however, be easily bypassed by renaming the frida-server binary, so we should find a better method.
+Starting with Android Nougat (API Level 24) the `ps` command will only return processes started by the user itself. When executing `ps` it will read the information from `/proc` and it's not possible to access information that belongs to other user ids.
+
+![Executing ps on Android Lollipop](Images/Chapters/0x05j/Android_Lollipop_ps.png)
+
+![Executing ps on Android Nougat](Images/Chapters/0x05j/Android_Nougat_ps.png)
+
+Even if the process name could easily be detected, this would only work if Frida is run in its default configuration. Perhaps it's also enough to stump some script kiddies during their first steps in reverse engineering. It can, however, be easily bypassed by renaming the frida-server binary. So because of this and the technical limitations of querying the process names in recent Android versions, we should find a better method.

```

Comment:
So because of this and the technical limitations:
Therefore, we need to look for a better method of querying the process names in recent android versions.

Reply:
I dont know of any to be honest. If it's a non-rooted device it seems you cannot query for  process names, only the one that matches with your uid.

Reply:
ah, then let's change it to:"
Therefore, we need to look for a better methods _than_ querying the process names in recent android versions.

Reply:
done